### ISSUE TITLE
[RFC] Multi-LoRA backend compatible with the Tinker API[/-]

### ISSUE THREAD

**Issue by yushengsu-thu** [radixark/miles#2258]
Multi-LoRA v2: a Tinker-protocol gateway over miles. Stacked PRs #2839–#2845, #2849; aggregate diff #2846 (do not merge). Supersedes the #2273/#2346/#2365 stack. Requires sglang [#38165](https://github.com/sgl-project/sglang/pull/38165) on `sglang-miles` and `tinker>=0.26.2` on the serving node.

### What the Tinker protocol requires of the server

Four orthogonal parts:

1. **Command log**: per model, a `seq_id`-ordered stream of `forward_backward` / `optim_step` / `save_state` / `load_state` / `save_weights_for_sampler`, executed in order, idempotent on retry;
2. **Training state**: (LoRA weights, Adam moments, accumulated grads) — `forward_backward` accumulates gradients, `optim_step` consumes them and advances the weights;
3. **Checkpoints**: `tinker://run/weights/name` paths are immutable once written;
4. **Sampling**: stateless read-only inference against a saved version.

The training loop is not on the server — it runs in the user's process, which also owns data, env, and reward.

### The gateway replaces the driver

miles' driver (`train_async.py`, ~200 lines of asyncio) hardcodes what happens next; every Tinker command corresponds exactly to one `await` in that loop. The difference between Tinker and miles is not in the execution layers — it is who owns the control flow. So the smallest implementation replaces the driver with a consumer of the command queue: TrainerController→Megatron and InferenceController→SGLang stay unchanged, and the data path (data source / rollout_fn / reward) is never instantiated.

### Scheduling: the real ordering constraint is weaker than seq_id

Consecutive `forward_backward`s between two `optim_step`s can run in any order — gradient accumulation is additive (the SDK itself sends chunks out of order). So each model's stream compiles to

```
stream(model) = [ batch of datums (unordered, per loss_fn) | barrier (optim/save/load/save_for_sampler) ]*
```

free within a batch, ordered at barriers, no constraints across models. That is the entire ordering requirement; everything else is scheduling policy. The code splits along the same boundary:

```
Stream   (semantics, per model)  seq_id dedup/order → batch|barrier structure
Planner  (policy, global)        pack ready datums across tenants → BatchUnit / BarrierUnit
Executor (mechanics, per slot)   tenant-blind execution on the trainer
```

The trainer is one SPMD collective domain, so the gateway issues from exactly one point: units are totally ordered before fan-out, one in flight at a time.

### Engine weight writes

A Tinker `SamplingClient` is fixed to the version that created it — an old sampler must keep answering with the old weights, because importance sampling needs to know which policy produced a sample. miles' `update_weights` overwrites in place, so every sampler sees the new weights immediately. A write needs a pause only when its target has readers:

- **in-place** (base sync, fixed-name single LoRA): readers exist → pause, stream, resume;
- **staged** (each `save_weights_for_sampler` writes to a new engine-side name, `run@7`): a new name has no readers → no pause. The engine registers the name as `defer_publish` (not servable while its bytes stream in) and makes it servable only after verifying a full checksum manifest — a lost bucket aborts the session and the name never serves.

Code branches on `if staged:`, never on a multi-LoRA feature flag; single LoRA is the same code path with no staged names.

Version residency follows three rules:

1. **Disk decides existence**: a version exists iff its checkpoint is on disk (that write is where `save_weights_for_sampler` commits); the engine's adapter registry is a cache that can be dropped and rebuilt;
2. **Sample requests are self-contained**: every request carries `(lora_name, path)`, so an engine that evicted the version reloads it from disk by itself — any saved version stays sampleable, including after the training lease expires;
3. **Creating a version and reading one are separate operations**: `save_weights_for_sampler` always creates a new version; sampling an old path requires no server action — the protocol has no load verb.

## Stages

**Stage 0 — system backbone**

| PR | Owns |
|---|---|
| #2839 | Remove multi-LoRA v1. |
| #2840 | Slot executor: MultiLoRALinear slots, per-slot decoupled Adam. |
| #2841 | Per-batch loss selection; the five Tinker losses. |
| #2842 | Per-slot checkpoints: weights, optimizer state, FP32 masters. |
| #2843 | `EngineWeightUpdateSession`; push versioned adapters through the weight-update stream. |
| #2844 | Gateway protocol layer: service/streams/planner/futures, API-key tenants, wire codecs. |
| #2845 | Runtime wiring and `serve_tinker.py`. |
| #2849 | Official-SDK example + gateway test suite. |

**Stage 1 — features & optimizations**

| PR | Owns |
|---|---|
| #3170 | Derive the slot capacity from measured memory. |

Candidates: LRU eviction pressure, per-tenant sampling quotas, `ttl_seconds`, multimodal inputs, agentic rollout, dynamic scaling.

**Comment by yushengsu-thu** [comment]
- check custom rollout func (should be able to use ---> figure this out
- fully async (use new version) ---> client side control (if use async --> I check the ans is Yes. - verified again

**Comment by yushengsu-thu** [comment]
- full fine-tune --> is it okay to use

**Comment by yushengsu-thu** [comment]
Tom split  (refer the Tom's PR) https://github.com/radixark/miles/pull/1842
RolloutManager:
- inference control
- rollout excutor
------

RolloutManager interface. The client directly generates and scores rollouts, builds Datum objects, and sends them to the backend trainer for forward_backward. ---> it will not use rolloutmanager

------
I would not write a separate, full RolloutManager for Tinker. Instead, reuse the split components with a Tinker-specific adapter:
- Reuse `InferenceController` to manage the SGLang fleet, router, engine lifecycle, and weight publication.
- Reuse `RolloutExecutor` as the training-batch scheduling container, but do not use its native dataset-driven rollout, reward, or data-conversion path. Plug in the Tinker operation-queue data source and TinkerRolloutFn, so it only pulls complete client-submitted Datum batches and passes them to the trainer.
This keeps sampling directly connected to the router, while the client continues to own verification and reward computation.

**Comment by yushengsu-thu** [comment]
comapre the rollout part: rolloutmanager, rolloutfunc in miles and in tinker.

**Comment by yushengsu-thu** [comment]
another point - need to be deal with the number of active lora > slots
megatron-bridge lora --> native lora (compatialbe but do not have this code)
agentic task rollout
